wordpress-mobile/WordPress-Android
# Issue/9560 offline featured image

Fixes #9560

Note: 
~**- This PR is dependant on https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1289 which needs to be merged first**~

Adds support for uploading post featured images asynchronously - without blocking the UI.

To test:
Initial steps for all test cases
1. Create a draft
2. Click on the overflow menu in the editor -> Post settings
3. Set a featured image from a local storage

--------------------

Successful upload
4. Notice the local image and a progress bar is shown during the upload
5. Notice the progress bar and it's overlay disappears when the upload finishes

--------------------

Failed upload with retry
4. Notice the local image and a progress bar is shown during the upload
5. Turn on airplane mode before the upload finishes
6. Notice the "tap for options" message
7. Turn off airplane mode
8. Click on the image and choose "Retry uploading"
9. Notice the local image and a progress bar is shown during the upload
10. Notice the progress bar and it's overlay disappears when the upload finishes

--------------------

Failed upload with remove
4. Notice the local image and a progress bar is shown during the upload
5. Turn on airplane mode before the upload finishes
6. Notice the "tap for options" message
7. Click on the image and choose "Remove"
8. Notice the image is removed, upload canceled and the error system notification dismissed

--------------------

Cancel ongoing upload
4. Notice the local image and a progress bar is shown during the upload
5. Click on the image and choose remove
6. Notice the image is removed, upload canceled and the error system notification dismissed

--------------------

Choose another image during ongoing upload
4. Notice the local image and a progress bar is shown during the upload
5. Click on the image and choose featured image option
6. Notice the previous image is removed, upload canceled and the error system notification dismissed
7. The newly picked image is being uploaded

--------------------

Post List label
4. Notice the local image and a progress bar is shown during the upload
5. Go back to the post list and notice "Uploading media" label is shown

--------------------

Remove all media dialog (test this for both Gutenberg and Aztec)
4. Notice the local image and a progress bar is shown during the upload
5. Turn on airplane mode
6. Go back to the editor and click on Update/Save/Publish
7. Publish confirmation dialog is shown -> click on Publish now
8. Notice another dialog is shown -> Click on "Remove all failed media"
9. The failed upload is removed and the error notification dismissed
10. Click on Update/Save/Publish again and notice the post gets published
(Note: It might be worth showing a snackbar message when we remove all the media - the flow seems a bit broken now, since nothing visible happens when you click on "Remove all failed media")


Update release notes:

- [ x ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.


Known issues:
~- If you cancel the upload when the "error uploading media" notification is shown, the notification doesn't get canceled.~
- We don't track any events - https://github.com/wordpress-mobile/WordPress-Android/issues/10151


Before            |  After
:-------------------------:|:-------------------------:
![featured-image](https://user-images.githubusercontent.com/2261188/60654406-3ecf8000-9e4c-11e9-9764-39c39cc2ecdd.gif) |  ![fixed-featured-image](https://user-images.githubusercontent.com/2261188/60654150-b7820c80-9e4b-11e9-9eb7-601fb132e904.gif)
